src/libmerklecpp/merklecpp.h

```diff
@@ -0,0 +1,1294 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <functional>
+#include <list>
+#include <memory>
+#include <sstream>
+#include <stack>
+#include <vector>
+
+#ifdef WIN32
+#  include <stdlib.h>
+#  if BYTE_ORDER == LITTLE_ENDIAN
+#    define htobe32(X) _byteswap_ulong(X)
+#    define be32toh(X) _byteswap_ulong(X)
+#  else
+#    define htobe32(X) (X)
+#    define be32toh(X) (X)
+#  endif
+#  undef max
+#endif
+
+#ifdef WITH_TRACE
+#  include <iostream>
+#  define TRACE_HASH_SIZE 3
+#  define TRACE(X) \
+    { \
+      X; \
+      std::cout.flush(); \
+    };
+#else
+#  define TRACE(X)
+#endif
+namespace merkle
+{
+  void serialise_size_t(size_t size, std::vector<uint8_t>& bytes)
+  {
+    // fix endianness?
+    uint8_t* size_bytes = (uint8_t*)&size;
+    for (size_t i = 0; i < sizeof(size_t); i++)
+      bytes.push_back(size_bytes[i]);
+  }
+
+  size_t deserialise_size_t(const std::vector<uint8_t>& bytes, size_t& index)
+  {
+    // fix endianness?
+    size_t result = 0;
+    uint8_t* result_bytes = (uint8_t*)&result;
+    for (size_t i = 0; i < sizeof(size_t); i++)
+      result_bytes[i] = bytes[index++];
+    return result;
+  }
+
+  template <size_t SIZE>
+  struct HashT
+  {
+    uint8_t bytes[SIZE];
+
+    HashT<SIZE>()
+    {
+      std::fill(bytes, bytes + SIZE, 0);
+    }
+
+    HashT<SIZE>(const uint8_t* bytes)
+    {
+      std::copy(bytes, bytes + SIZE, this->bytes);
+    }
+
+    HashT<SIZE>(const std::string& s)
+    {
+      if (s.length() != 2 * SIZE)
+        throw std::runtime_error("invalid hash string");
+      for (size_t i = 0; i < SIZE; i++)
+      {
+        int tmp;
+        sscanf(s.c_str() + 2 * i, "%02x", &tmp);
+        bytes[i] = tmp;
+      }
+    }
+
+    HashT<SIZE>(const std::vector<uint8_t>& bytes)
+    {
+      if (bytes.size() < SIZE)
+        throw std::runtime_error("not enough bytes");
+      deserialise(bytes);
+    }
+
+    HashT<SIZE>(const std::vector<uint8_t>& bytes, size_t& position)
+    {
+      if (bytes.size() - position < SIZE)
+        throw std::runtime_error("not enough bytes");
+      deserialise(bytes, position);
+    }
+
+    HashT<SIZE>(const std::array<uint8_t, SIZE>& bytes)
+    {
+      std::copy(bytes.data(), bytes.data() + SIZE, this->bytes);
+    }
+
+    size_t size() const
+    {
+      return SIZE;
+    }
+
+    size_t byte_size() const
+    {
+      return SIZE;
+    }
+
+    std::string to_string(size_t num_bytes = SIZE) const
```

Comment:
Should this be using `fmt::format` for consistency with other hash code, e.g. https://github.com/microsoft/CCF/blob/master/src/crypto/hash.h#L26-L40?

Reply:
That should be added in a separate file with the LOG_ macros that are specific to CCF. That way if/when this gets spun off into its own header-only library in a separate repo, it won't have a dependency on fmtlib just for this one thing.